lib/core/src/display/world.rs

```diff
@@ -28,221 +42,262 @@ use eval_tt::*;
+            panel.begin();
+        }
+    }
+
+    fn end(&mut self) {
+        for panel in &self.panels {
+            panel.end();
+        }
+        self.monitor.draw();
+        self.stats.reset();
+    }
+}
+
+
+
+// =================
+// === WorldData ===
+// =================
+
+// === Definition ===
 
 /// World is the top-level structure managing several instances of `Workspace`.
 /// It is responsible for updating the system on every animation frame.
 #[derive(Derivative)]
 #[derivative(Debug(bound=""))]
-pub struct World {
-    pub workspaces      : OptVec<Workspace>,
+pub struct WorldData {
+    pub workspace       : Workspace,
     pub workspace_dirty : WorkspaceDirty,
     pub logger          : Logger,
     pub event_loop      : EventLoop,
+    pub variables       : UniformScope,
+    pub performance     : Performance,
+    pub start_time      : f32,
+    pub time            : Uniform<f32>,
+    pub display_mode    : Uniform<i32>,
     pub fonts           : Fonts,
     pub update_handle   : Option<CallbackHandle>,
-    pub self_reference  : Option<WorldRef>,
+    pub stats           : Stats,
+    pub stats_monitor   : StatsMonitor,
 }
 
 
 // === Types ===
 
 pub type WorkspaceID    = usize;
-pub type WorkspaceDirty = dirty::SharedSet<WorkspaceID>;
+pub type WorkspaceDirty = dirty::SharedBool;
 promote_workspace_types!{ [[WorkspaceOnChange]] workspace }
 
 
 // === Callbacks ===
 
 closure! {
-fn workspace_on_change(dirty:WorkspaceDirty, ix:WorkspaceID) -> WorkspaceOnChange {
-    || dirty.set(ix)
+fn workspace_on_change(dirty:WorkspaceDirty) -> WorkspaceOnChange {
+    || dirty.set()
 }}
 
 
 // === Implementation ===
 
-impl World {
+impl WorldData {
     /// Create and initialize new world instance.
     #[allow(clippy::new_ret_no_self)]
-    pub fn new() -> WorldRef {
+    pub fn new<Dom:Str>(dom:Dom) -> World {
         println!("NOTICE! When profiling in Chrome check 'Disable JavaScript Samples' under the \
                   gear icon in the 'Performance' tab. It can drastically slow the rendering.");
-        let world_ref  = WorldRef::new(Self::new_uninitialized());
-        let world_ref2 = world_ref.clone_rc();
-        let world_ref3 = world_ref.clone_rc();
-        with(world_ref.borrow_mut(), |mut data| {
-            let update          = move || world_ref2.borrow_mut().run();
+        let world     = World::new(Self::new_uninitialized(dom));
+        let world_ref = world.clone_ref();
+        with(world.borrow_mut(), |mut data| {
+            let update          = move || world_ref.borrow_mut().run();
             let update_handle   = data.event_loop.add_callback(update);
             data.update_handle  = Some(update_handle);
-            data.self_reference = Some(world_ref3);
         });
-        world_ref
+
+        // -----------------------------------------------------------------------------------------
+        // FIXME: Hacky way of switching display_mode. To be fixed and refactored out.
```

Comment:
You should add [WD] marker to fixme, according to code style guide

Reply:
Oh, totally forgot about it